pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/NoDictionarySingleColumnGroupKeyGenerator.java

```diff
@@ -378,22 +486,30 @@ private int getKeyForValue(ByteArray value) {
   private static class IntGroupKeyIterator implements Iterator<GroupKey> {
     final Iterator<Int2IntMap.Entry> _iterator;
     final GroupKey _groupKey;
+    Integer _groupKeyForNullValue;
 
-    IntGroupKeyIterator(Int2IntOpenHashMap intMap) {
+    IntGroupKeyIterator(Int2IntOpenHashMap intMap, Integer groupKeyForNullValue) {
       _iterator = intMap.int2IntEntrySet().fastIterator();
       _groupKey = new GroupKey();
+      _groupKeyForNullValue = groupKeyForNullValue;
     }
 
     @Override
     public boolean hasNext() {
-      return _iterator.hasNext();
+      return _iterator.hasNext() || _groupKeyForNullValue != null;
     }
 
     @Override
     public GroupKey next() {
-      Int2IntMap.Entry entry = _iterator.next();
-      _groupKey._groupId = entry.getIntValue();
-      _groupKey._keys = new Object[]{entry.getIntKey()};
+      if (_iterator.hasNext()) {
+        Int2IntMap.Entry entry = _iterator.next();
+        _groupKey._groupId = entry.getIntValue();
+        _groupKey._keys = new Object[]{entry.getIntKey()};
+      } else if (_groupKeyForNullValue != null) {
+        _groupKey._groupId = _groupKeyForNullValue;
+        _groupKey._keys = new Object[]{null};
+        _groupKeyForNullValue = null;
+      }
```

Comment:
Consider returning `null` group first to reduce the overhead, also add a TODO to revisit this because this one adds overhead to regular case without null handling
```suggestion
      if (_groupKeyForNullValue != null) {
        _groupKey._groupId = _groupKeyForNullValue;
        _groupKey._keys = new Object[]{null};
        _groupKeyForNullValue = null;
        return _groupKey;
      }
      Int2IntMap.Entry entry = _iterator.next();
      _groupKey._groupId = entry.getIntValue();
      _groupKey._keys = new Object[]{entry.getIntKey()};
```